src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs

```diff
@@ -0,0 +1,424 @@
+
+            var noSsl2 = (options & libssl.Options.SSL_OP_NO_SSLv2) != 0;
+            var noSsl3 = (options & libssl.Options.SSL_OP_NO_SSLv3) != 0;
+            var noTls10 = (options & libssl.Options.SSL_OP_NO_TLSv1) != 0;
+            var noTls11 = (options & libssl.Options.SSL_OP_NO_TLSv1_1) != 0;
+            var noTls12 = (options & libssl.Options.SSL_OP_NO_TLSv1_2) != 0;
+
+            var method = IntPtr.Zero;
+            if (noSsl2 && noSsl3 && noTls11 && noTls12)
+            {
+                method = isServer ? libssl.TLSv1_server_method() : libssl.TLSv1_client_method();
+            }
+            else if (noSsl2 && noSsl3 && noTls10 && noTls12)
+            {
+                method = isServer ? libssl.TLSv1_1_server_method() : libssl.TLSv1_1_client_method();
+            }
+            else if (noSsl2 && noSsl3 && noTls10 && noTls11)
+            {
+                method = isServer ? libssl.TLSv1_2_server_method() : libssl.TLSv1_2_client_method();
+            }
+            else if (noSsl2 && noTls10 && noTls11 && noTls12)
+            {
+                method = isServer ? libssl.SSLv3_server_method() : libssl.SSLv3_client_method();
+            }
+            else
+            {
+                method = isServer ? libssl.SSLv23_server_method() : libssl.SSLv23_client_method();
+            }
+
+            if (IntPtr.Zero == method)
+            {
+                throw CreateSslException("Failed to get SSL method");
+            }
+
+            return method;
+        }
+
+        //TODO See if SSL_CTX_Set_quite_shutdown can be used
+        private static void Disconnect(IntPtr sslPtr)
+        {
+            if (IntPtr.Zero != sslPtr)
+            {
+                int retVal = libssl.SSL_shutdown(sslPtr);
+                if (0 == retVal)
+                {
+                    retVal = libssl.SSL_shutdown(sslPtr);
+                    if (retVal < 0)
+                    {
+                        libssl.SSL_get_error(sslPtr, retVal);
+                    }
+                }
+
+                libssl.SSL_free(sslPtr);
+            }         
+        }
+
+        //TODO should we check Bio should retry?
+        private static int BioRead(IntPtr BioPtr, IntPtr buffer, int count)
+        {
+            int bytes = libssl.BIO_read(BioPtr, buffer, count);
+            if (bytes != count)
+            {
+                throw CreateSslException("Failed in Read BIO");
+            }
+            return bytes;
+        }
+
+        //TODO should we check Bio should retry?
+        private static int BioWrite(IntPtr BioPtr, IntPtr buffer, int count)
+        {
+            int bytes = libssl.BIO_write(BioPtr, buffer, count);
+            if (bytes != count)
+            {
+                throw CreateSslException("Failed in Write BIO");
+            }
+            return bytes;
+        }
+
+        private static int GetSslError(IntPtr sslPtr, int result, string message)
```

Comment:
Removed unused message parameter. CreateSslException is called based on output from GetSslError at one place. Since CreateSslException is a generic function, do not want to add libssl.SslErrorCode.SSL_ERROR_WANT_READ check in CreateSslException method. So keeping GetSslError as a seperate method.